### PR TITLE
dlopen test

### PR DESCRIPTION
**Pull request by gilles-peskine-arm**
Test that `dlopen("libmbedtls.so")` works. Non-regression test for https://github.com/ARMmbed/mbedtls/pull/5126.

### CODE REVIEW

File: programs/test/CMakeLists.txt [hunk 1]
```diff
@@ -27,6 +27,15 @@ if(TEST_CPP)
     target_link_libraries(cpp_dummy_build ${mbedcrypto_target})
 endif()
 
+if(USE_SHARED_MBEDTLS_LIBRARY AND
+   NOT ${CMAKE_SYSTEM_NAME} MATCHES "[Ww][Ii][Nn]")
+    add_executable(dlopen "dlopen.c")
+    target_include_directories(dlopen PRIVATE ${CMAKE_CURRENT_SOURCE_DIR}/../../include)
+    if(${CMAKE_SYSTEM_NAME} MATCHES "Linux")
+        target_link_libraries(dlopen "-ldl")
```
Comment (haampie) [inline programs/test/CMakeLists.txt:35]:
```suggestion
        target_link_libraries(dlopen ${CMAKE_DL_LIBS})
```

File: programs/test/dlopen.c [hunk 2]
```diff
@@ -0,0 +1,102 @@
...
+            fprintf( stderr, "Dynamic loading error for %s(%s): %s\n",  \
+                     function, argument, CHECK_DLERROR_error );         \
+            mbedtls_exit( MBEDTLS_EXIT_FAILURE );                       \
+        }                                                               \
+    }                                                                   \
+    while( 0 )
+
+int main( void )
+{
+    unsigned n;
```
Comment (tom-cosgrove-arm) [inline programs/test/dlopen.c:58]: Should be guarded by `#if defined(MBEDTLS_SSL_TLS_C) || defined(MBEDTLS_MD_C)`

**Comment by tom-cosgrove-arm** [comment]
Since #5126 is merged, this can be rebased

**Comment by gilles-peskine-arm** [comment]
I've rebased on top of `development`. The previous version is in https://github.com/gilles-peskine-arm/mbedtls/tree/dlopen-test-1 . I'll push new commits to address the review comments.

File: programs/test/CMakeLists.txt (hunk 1 above)
Comment (gilles-peskine-arm) [reply programs/test/CMakeLists.txt]: Thanks, I'd missed that.

It's not for Windows which doesn't have dlopen anyway. It's for non-Linux POSIX systems such as *BSD.

File: programs/test/dlopen_demo.sh [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+fi
+export LD_LIBRARY_PATH
+
+# OSX/macOS
+if [ -n "${DYLD_LIBRARY_PATH-}" ]; then
+    DYLD_LIBRARY_PATH="$library_dir:$DYLD_LIBRARY_PATH"
+else
+    DYLD_LIBRARY_PATH="$library_dir"
+fi
+export DYLD_LIBRARY_PATH
```
Comment (tom-cosgrove-arm) [inline programs/test/dlopen_demo.sh:42]: How should I test this on macOS?

I get

```
mini1 $ git show | head -11
commit 7fb54c567456f3501ed711bbb72d1d2de2d130da
Author: Gilles Peskine <Gilles.Peskine@arm.com>
Date:   Wed Nov 10 21:04:24 2021 +0100

    More explicit output for the test program

    Without that, the logs were a bit hard to understand if you didn't know what
    to expect.

    Signed-off-by: Gilles Peskine <Gilles.Peskine@arm.com>

mini1 $ ls
3rdparty/		ChangeLog		Makefile.all.bak	configs/		library/
BRANCHES.md		ChangeLog.d/		README.md		dco.txt			programs/
BUGS.md			DartConfiguration.tcl	SECURITY.md		docs/			scripts/
CMakeLists.txt		LICENSE			SUPPORT.md		doxygen/		tests/
CONTRIBUTING.md		Makefile		cmake/			include/		visualc/

mini1 $ find . -name libmbedtls\*
./library/libmbedtls.dylib
./library/libmbedtls.a

mini1 $ sh programs/test/dlopen_demo.sh
Running dynamic loading test program: programs/test/dlopen
Loading libraries from: programs/test/../../library
Dynamic loading error for dlopen(libmbedtls.so): dlopen(libmbedtls.so, 2): image not found
```

Comment (gilles-peskine-arm) [reply programs/test/dlopen_demo.sh:42]: You tested it right. I'd only done half of the work required to support macOS. Fixed.

**Review by tom-cosgrove-arm — APPROVED** [review]
LGTM

**Comment by haampie** [comment]
Can I suggest using https://github.com/vtjnash/libwhich, it works on Windows too, and this is actually how I initially found the problem with mbedtls. You'd use it like this:

```
$ libwhich /absolute/path/to/libwhich.so
failed to open library: /home/user/spack/opt/spack/linux-ubuntu20.04-skylake/gcc-9.3.0/mbedtls-3.0.0-cbg7xw3mbyxkrbf6hfngw6xgdxchxgzp/lib/libmbedtls.so.16: undefined symbol: mbedtls_x509_crt_profile_default
$ echo $?
1
```

**Comment by tom-cosgrove-arm** [comment]
@haampie @gilles-peskine-arm Re `libwhich` yes, it would allow us to do the same thing, but for mbedTLS it is better to have this as just part of the tests, without any external dependencies

**Review by davidhorstmann-arm — CHANGES_REQUESTED** [review]
Looks good apart from a small quibble for out-of-tree builds

File: programs/test/dlopen_demo.sh [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+# WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+set -e -u
+
+program_dir="${0%/*}"
+program="$program_dir/dlopen"
```
[the remark this answers is not in the document]
Comment (davidhorstmann-arm) [reply programs/test/dlopen_demo.sh:24]: Alternatively, I think you could use `link_to_source()` in `CMakeLists.txt`, which would probably be much easier.

[the remark this answers is not in the document]
Comment (gilles-peskine-arm) [reply programs/test/dlopen_demo.sh:24]: I guess the same applies to `key_ladder_demo.sh`?

Would you mind proposing a patch?

[the remark this answers is not in the document]
Comment (davidhorstmann-arm) [reply programs/test/dlopen_demo.sh:24]: I'll approve this one in that case, and make a new PR.

**Review by davidhorstmann-arm — APPROVED** [review]

File: programs/test/dlopen_demo.sh (hunk 4 above)
[the remark this answers is not in the document]
Comment (gilles-peskine-arm) [reply programs/test/dlopen_demo.sh:24]: FYI testing the sample programs is on our backlog. I can't find a task on GitHub but it's definitely in our mental backlog and may be in the Jira we no longer use. There is https://github.com/ARMmbed/mbedtls/pull/2698 which adds a second `_demo.sh` script and a `run_demos.py`. It doesn't support out-of-tree cmake builds however, but I think that's largely orthogonal.

**Comment by tom-cosgrove-arm** [comment]
@gilles-peskine-arm Would you like me to prepare a backport to 2.x?

File: programs/test/dlopen_demo.sh (hunk 4 above)
[the remark this answers is not in the document]
Comment (davidhorstmann-arm) [reply programs/test/dlopen_demo.sh:24]: The PR making this change is at #5237

**Pull request merged**